server/src/main/scala/org/http4s/server/ManagedServerApp.scala

```diff
@@ -0,0 +1,26 @@
+package org.http4s
+package server
+
+import scalaz.{\/, Codensity, ImmutableArray}
+import scalaz.concurrent.Task
+
+import org.http4s.util.managed.ManagedApp
+
+trait ManagedServerApp extends ManagedApp {
+  final def run(args: Vector[String]): Codensity[Task, Unit] =
+    runServer(args).flatMap { server =>
+      Codensity.rep(terminateServer(server))
+    }
+
+  /** Prompts for input on the console, and then terminates the server.
+    * Override to implement your own termination condition.
+    */
+  def terminateServer(server: Server): Task[Unit] =
+    Task.async { cb =>
+      cb(\/.fromTryCatchNonFatal {
+        Console.readLine(s"Press <Enter> to terminate server on ${server.address}\n")
```

Comment:
Note: changing this to a JVM shutdown hook for ^C doesn't wind down the stack the way that this implementation does.  I think the thread on which the codensity runs already exits.  There might be a way to salvage this with a daemon thread and a lock, but this is where I start to feel seasick.